VERSIONING.md

```diff
@@ -0,0 +1,157 @@
+   1. [Areas that are *NOT* covered](#areas-that-are-not-covered)
+1. [FAQ](#faq)
+   1. [How often is Vector released?](#how-often-is-vector-released)
+   1. [How does Vector treat patch and minor versions?](#how-does-vector-treat-patch-and-minor-versions)
+   1. [How does Vector treat major versions \(breaking changes\)?](#how-does-vector-treat-major-versions-breaking-changes)
+   1. [How does Vector treat pre-1.0 versions?](#how-does-vector-treat-pre-10-versions)
+   1. [Release cadence](#release-cadence)
+
+<!-- /MarkdownTOC -->
+
+## Convention
+
+Vector adheres to the [Semantic Versioning 2.0] convention. In summary:
+
+* Versions follow the `MAJOR.MINOR.PATCH` format (i.e., `2.5.1`)
+* `PATCH` increments only when backward compatible bug fixes are introduced.
+* `MINOR` increments only when new, backward compatible functionality is introduced.
+* `MAJOR` increments if any backwards incompatible changes are introduced.
+* Pre `1.0` (major version `0`) is for initial development and `MINOR` version bumps can introduce breaking changes.
+
+## Public API
+
+Semantic Versioning hinges on Vector's defintion of "public API". By nature of
+Vector - a tool that collects, processes, and routes data from disparate systems
+- it has a very large public surface area. It's not immediately obvious which
+parts are covered under our versioning contract and how they're covered. This
+section aims to remove all ambiguity in this area.
+
+### Areas that *are* covered
+
+The following Vector areas are covered in Vector's definition of public API.
+
+#### Intended for *public* consumption
+
+The follow Vector areas are inteded for *public* consumption (consumption by
+anything other than Vector itself). Backward incompatble changes will trigger
+a major version increment.
+
+* [CLI]
+  * The root [`vector` command], its flags and exit code
+  * The [`vector validate` subcommand], the flags exit codes
+* [Data model]
+  * As exposed in the source of the [`lua` transform]
+* [GraphQL API]
+* Telemetry
+  * Vector's internal metrics as provided by the [`internal_metrics` source]
+* [VRL]
+
+#### Intended for *private* consumption
+
+The following Vector areas are inteded for *private* consumption (consumption by
+Vector only). Backward incompatble changes will trigger a major version
+increment only if Vector itself is not compatbile with previous versions.
+
+* [Configuration schema]
+* [Data directory] and its contents
+* [Data model]
+  * As exposed in the output of the [`vector` sink]
+
+### Areas that are *NOT* covered
+
+The following Vector areas are *not* covered in Vector's definition of Public
+API. Breaking changes in these areas will *not* trigger a major version
+increment.
+
+* [CLI]
+  * The [`vector generate` subcommand]
+  * The [`vector graph` subcommand]
+  * The [`vector help` subcommand]
+  * The [`vector list` subcommand]
+  * The [`vector tap` subcommand]
+  * The [`vector top` subcommand]
+  * The [`vector vrl` subcommand]
```

Comment:
I would just simplify this to "command inputs and outputs" since we don't currently make any guarantees about the input/output of any command.